src/Core/MySQL/PacketsProtocolBinary.cpp

```diff
@@ -0,0 +1,363 @@
+#include <Columns/IColumn.h>
+#include <Core/MySQL/IMySQLReadPacket.h>
+#include <Core/MySQL/IMySQLWritePacket.h>
+#include <Core/MySQL/PacketsProtocolBinary.h>
+#include "Common/LocalDate.h"
+#include "Common/LocalDateTime.h"
+#include "Columns/ColumnLowCardinality.h"
+#include "Columns/ColumnVector.h"
+#include "Columns/ColumnsDateTime.h"
+#include "Core/DecimalFunctions.h"
+#include "DataTypes/DataTypeDateTime64.h"
+#include "DataTypes/DataTypeLowCardinality.h"
+#include "DataTypes/DataTypeNullable.h"
+#include "DataTypes/DataTypesNumber.h"
+#include "Formats/FormatSettings.h"
+#include "IO/WriteBufferFromString.h"
+#include "base/DayNum.h"
+#include "base/Decimal.h"
+#include "base/types.h"
+
+namespace DB
+{
+
+namespace MySQLProtocol
+{
+
+    namespace ProtocolBinary
+    {
+        ResultSetRow::ResultSetRow(
+            const Serializations & serializations_, const DataTypes & data_types_, const Columns & columns_, int row_num_)
+            : row_num(row_num_), columns(columns_), data_types(data_types_), serializations(serializations_)
+        {
+            /// See https://dev.mysql.com/doc/dev/mysql-server/8.1.0/page_protocol_binary_resultset.html#sect_protocol_binary_resultset_row
+            payload_size = 1 + null_bitmap_size;
+            FormatSettings format_settings;
+            for (size_t i = 0; i < columns.size(); ++i)
+            {
+                ColumnPtr col = columns[i];
+                if (col->isNullAt(row_num))
+                {
+                    size_t byte = (i + 2) / 8;
+                    int bit = 1 << ((i + 2) % 8);
+                    null_bitmap[byte] |= bit;
+                    continue; // NULLs are stored in the null bitmap only
+                }
+
+                DataTypePtr data_type = removeLowCardinality(removeNullable((data_types[i])));
+                TypeIndex type_index = data_type->getTypeId();
+                switch (type_index)
+                {
+                    case TypeIndex::Int8:
+                        payload_size += 1;
+                        break;
+                    case TypeIndex::UInt8:
+                        if (data_type->getName() == "Bool")
+                        {
```

Comment:
(see my other comment about `BIT` vs. `BOOL`)

Reply:
Isn't this one coming from ClickHouse itself?